### DISCUSSION TITLE
Confused about the actual use case

### DISCUSSION THREAD

**Discussion by emrakyz** [pemistahl/grex#264] [Q&A]
First of all, thanks a lot for this tool.

The idea looks really good and promising but I couldn't understand what the actual use case was. The documentation lacks in terms of various interesting examples.

I have tried to refind some of the patterns I had already written myself just to test `grex`.

For example I use the below pattern to extract doi addresses from various inputs and/or pdfs:
`sed -n -E 's/.*((DOI|doi)((\.(org))?\/?|:? *))([^: ]+[^ .]).*/doi:\6/p; q'`

The actual pattern is this:
`.*((DOI|doi)((\.(org))?\/?|:? *))([^: ]+[^ .]).*`

The part that I aim to capture is (6th captured group which is the DOI Address): `([^: ]+[^ .])`

As an example, it captures this part at the end: `10.36227/techrxiv.22659061.v1`

I have tried to place lots of valid cases on each line (doi addresses in different forms as in the above regex pattern) to a `test.txt` file.

I used `grex -r -g -c --no-start-anchor -f "test.txt"` command. I knew that it couldn't give me a pattern similar to my original one but the resulting output was even much more different than I expected. I got an extremely long regex pattern which also captured unwanted parts (false-positive constants) that would break the command for my actual use case. This is understandable but impossible to avoid without infinite examples that are completely different from each other, except the actual constants.

I have also tried to test different cases in order to refind the regex patterns I had written before with simpler patterns using some made-up examples in the test file. The below pattern that had been written before, could be an example: 
`^ *([0-9]+).*\s{2,}(.+)$`

But `grex` outputs a pattern that is always wrong and very long; not similar to my actual pattern. The output is not "wrong" in technical sense but definitely not usable to achieve something. It's not even appropriate to be modified to some extent manually and then used. No matter the sample size, this was always the case in my tests.

Even for very basic cases; since we can't be expressive enough, the output is not usable. Without proper expression, this tool can only create patterns which are usable only with almost infinite example cases that cover all possibilities.

The problem is that, - as far as I understand - we can't be expressive enough especially in terms of constants, variables, wanted parts, unwanted parts, the actual main pattern that should captured and all. Without these, I could not find a proper use case but I really want to use this tool in an actual scenario. How can we be more expressive so we can automate creating at least a base pattern to work on?

I have also tried to find the final pattern in a segmented way but failed similarly.

Writing regex is fairly easy for small, simple tasks. What I initially had in mind for this tool was that it would be helpful for us to create regex, for very complex patterns easily in a more efficient, more correct way. Right now I feel like we have a very powerful and robust but useless tool. Is this just "experimental" or a kind of a base that will be used by future tools?

Could you please inform what I do wrong? What is the best practice to solve an actual problem using `grex`? What type of problems are best to get help from `grex`?

I probably misunderstood the tool or made a mistake regarding the intended use case.

**Comment by pemistahl** [comment]
I'm sorry for the late response. Well, the intended use case currently is just what the README says:

> The currently best use case for *grex* is to find an initial correct regex which should be inspected by hand if further optimizations are possible.

The programming involved in the current state of the tool is already very complex. It is desirable to make the tool more powerful by allowing more generalizations. But this is hard to implement and it's difficult to guarantee valid and correct output for each arbitrary input. 

Feel free to improve the implementation and make the tool more intelligent. PRs are always welcome.

**Reply by emrakyz** [reply]
Dear @pemistahl, I totally understand but it does not give you an "initial correct regex" to work on. It simply captures everything, and in the below image, I tried to simplify the test case. If I add actual `pdftotext` outputs there, the output will be even more terrible. 

We do not use regex to capture all of the possible entries. Even if I add a billion entries in that text file to increase the sample rate; there will be infinite numbers of more possibilities and it will still match false positives and in that case, I believe the generated regex would be millions of characters long.

It simply pastes almost every character it sees in the text file.

For example, I had written this regex to grab everything between "DOI" or "doi" (including optional ".org" or colon variations) and the next space or period, while ignoring any text before or after it. The logic behind regex is to abstract everything because we want to match the parts that will be unknown to us and eliminate parts that are definitely known beforehand; as well as to add some parts that could be possibilities.

The tool produces overly specific and lengthy patterns that capture literal text rather than meaningful abstractions. This is not just an optimization problem. The output would never ever work in a real life scenario. There's no way to indicate:

- Which parts should be constants vs. variables
- What parts are important to capture
- What should be explicitly excluded
- The semantic structure of the pattern

Without semantic hints, it can't distinguish between incidental text and meaningful patterns (like recognizing what makes a valid DOI).

![image](https://github.com/user-attachments/assets/023f02ae-4a0b-40f4-af26-857579812bdf)

Here is an extremely simpler one. This is not a regex you would want to use in any form or shape:

![image](https://github.com/user-attachments/assets/415c9313-f7a2-4a26-aeb4-c4110f3218d6)

It has a clear pattern to humans:

- Role prefix (user/admin/mod)
- Numbers of varying length
- Status suffix in caps
- Underscore separator

Instead it,
- Treats each number as a separate literal case
- Fails to properly generalize the number patterns
- May not properly group the prefixes and suffixes
- Could even treat entire strings as alternations

This tool's problem is not that it requires further optimizations to begin with. It requires you to create a regex from scratch. Even if I create a text file with extremely simple words such as `aha, chat, that, phat`; the output from the tool would be even longer than the `test.txt` file.

**The tool seems more like a developer exercise on Rust and regex algorithms.** In that case, I totally respect and I understand the already complex structure behind it; but I do not understand why would any person keep this program on their system and use it. I would actually be very glad if someone shares a real life usecase where they create a regex pattern they use anywhere.

The question being included in the readme, is extremely misleading even if it answers the question with "yes":

> Do I still need to learn to write regexes then?

As if the tool is powerful enough to create meaningful regexes but the only problem is some optimization.

I would definitely like to help but this is certainly beyond my expertise and I don't program in Rust. I hope these examples can encourage other people though.

Thanks for reading.
